libs/init_runtime/src/hpx_init.cpp

```diff
@@ -900,7 +900,7 @@ namespace hpx {
     }    // namespace detail
 
     ///////////////////////////////////////////////////////////////////////////
-    int finalize(double shutdown_timeout, double localwait, error_code& ec)
+    HPX_EXPORT int finalize(double shutdown_timeout, double localwait, error_code& ec)
```

Comment:
Shouldn't those API functions be marked `HPX_EXPORT` at their declaration instead?

Reply:
Yes... maybe. I need to dig a bit more. We only include the declarations for applications. I think I'll put some of the declarations in a separate header file (some are currently duplicated, some are not included in hpx_init.cpp).

Reply:
I just pushed something that works on Linux. It remains to be seen if it's correct on Windows.

Reply:
I'm starting to suspect that the linker errors weren't at all due to missing or misplaced `HPX_EXPORTS`, but due to the force linking stuff not working. @hkaiser I attempted to add something to fix it based on the other force linking files, but I don't know if it's enough. If this comes back red would you mind telling me exactly what I should add to make it link? :) Does everything that's exported from a module need a force linking helper?

Reply:
The native MSVC linker adds functions based on whether at least one other symbol from the same object file is needed (that's the default - however, it can do function level linking as well, but we don't enable this). That means that all symbols from an object file will be forced into the shared library if at least one of those symbols is needed. We use this to create an artificial reference to one of the symbols from files that would otherwise be left out (as the core library itself doesn't require linking to any of those symbols). That's what the `force_linking structures are all about` - creating that artificial reference.

Reply:
Please note also, that this is orthogonal to the export issue. Symbols will be included through the force-linking regardless whether they are to exported or not.

I general, I think this is a bug in the linker. I would expect for it to link in every exported symbol from static libraries, even if they are not directly referenced by the linked module itself..

Reply:
None of the additional `HPX_EXPORT`s were needed in the end but I reshuffled things a bit (added another header file with declarations, so that `hpx_init.cpp` and `hpx_init_impl.hpp` etc. can share the declarations). Since object libraries went on hold indefinitely in #4649, I fixed the force linking for `init_runtime` and `runtime_local` here. Thanks @hkaiser for the explanation! It was the part about one symbol per translation unit that I missed. I thought it was enough to force one symbol per library to be linked.